evennia/contrib/puzzles.py

```diff
@@ -0,0 +1,747 @@
+(aka results) are automatically created.
+
+Consider this simple Puzzle:
+
+    orange, mango, yogurt, blender = fruit smoothie
+
+As a Builder:
+
+    @create/drop orange
+    @create/drop mango
+    @create/drop yogurt
+    @create/drop blender
+    @create/drop fruit smoothie
+
+    @puzzle smoothie, orange, mango, yogurt, blender = fruit smoothie
+    ...
+    Puzzle smoothie(#1234) created successfuly.
+
+    @destroy/force orange, mango, yogurt, blender, fruit smoothie
+
+    @armpuzzle #1234
+    Part orange is spawned at ...
+    Part mango is spawned at ...
+    ....
+    Puzzle smoothie(#1234) has been armed successfully
+
+As Player:
+
+    use orange, mango, yogurt, blender
+    ...
+    Genius, you blended all fruits to create a fruit smoothie!
+
+Details:
+
+Puzzles are created from existing objects. The given
+objects are introspected to create prototypes for the
+puzzle parts and results. These prototypes become the
+puzzle recipe. (See PuzzleRecipe and @puzzle
+command). Once the recipe is created, all parts and result
+can be disposed (i.e. destroyed).
+
+At a later time, a Builder or a Script can arm the puzzle
+and spawn all puzzle parts in their respective
+locations (See @armpuzzle).
+
+A regular player can collect the puzzle parts and combine
+them (See use command). If player has specified
+all pieces, the puzzle is considered solved and all
+its puzzle parts are destroyed while the puzzle results
+are spawened on their corresponding location.
+
+Installation:
+
+Add the PuzzleSystemCmdSet to all players.
+Alternatively:
+
+    @py self.cmdset.add('evennia.contrib.puzzles.PuzzleSystemCmdSet')
+
+"""
+
+import itertools
+from random import choice
+from evennia import create_object, create_script
+from evennia import CmdSet
+from evennia import DefaultObject
+from evennia import DefaultScript
+from evennia import DefaultCharacter
+from evennia import DefaultRoom
+from evennia import DefaultExit
+from evennia.commands.default.muxcommand import MuxCommand
+from evennia.utils.utils import inherits_from
+from evennia.utils import search, utils, logger
+from evennia.prototypes.spawner import spawn
+
+# Tag used by puzzles
+_PUZZLES_TAG_CATEGORY = 'puzzles'
+_PUZZLES_TAG_RECIPE = 'puzzle_recipe'
+# puzzle part and puzzle result
+_PUZZLES_TAG_MEMBER = 'puzzle_member'
+
+_PUZZLE_DEFAULT_FAIL_USE_MESSAGE = 'You try to utilize %s but nothing happens ... something amiss?'
+_PUZZLE_DEFAULT_SUCCESS_USE_MESSAGE = 'You are a Genius!!!'
+_PUZZLE_DEFAULT_SUCCESS_USE_LOCATION_MESSAGE = "|c{caller}|n performs some kind of tribal dance and |y{result_names}|n seems to appear from thin air"
+
+# ----------- UTILITY FUNCTIONS ------------
+
+def proto_def(obj, with_tags=True):
+    """
+    Basic properties needed to spawn
+    and compare recipe with candidate part
+    """
+    protodef = {
+        # FIXME: Don't we need to honor ALL properties? attributes, contents, etc.
+        'prototype_key': '%s(%s)' % (obj.key, obj.dbref),
```

Comment:
Added `prototype_key` but tests still fail and it seems to be related to `tags`. 

See line 114 below and how I used to deal with `tags` in 0.7.0. Perhaps there is something else I need to do in 0.8.0?